### PR TITLE
[webgpu] Add gather operator

### PR DESCRIPTION
**Pull request by haoyunfeix**
FEATURE

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/4791)

### CODE REVIEW

**Comment by haoyunfeix** [comment]
@qjia7 @axinging @xhcao @gyagp PTAL

File: tfjs-backend-webgpu/src/kernels/gather_webgpu.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+  constructor(aShape: number[], outputShape: number[]) {
+    this.outputShape = aShape.slice();
+    this.aShape = aShape;
+    this.outputShape = outputShape;
+    this.rank = outputShape.length;
+    this.dispatchLayout = flatDispatchLayout(this.outputShape);
+    this.dispatch = computeDispatch(
+        this.dispatchLayout, this.outputShape, this.workGroupSize,
+        [this.workPerThread, 1, 1]);
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/kernels/gather_webgpu.ts:42]: `this.workPerThread` is not used in your shader. But you define it as a non-one value.

File: tfjs-backend-webgpu/src/kernels/gather_webgpu.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    this.shaderKey = `gather`;
+  }
+  getUserCode(): string {
+    const dtype = getCoordsDataType(this.rank);
+    const sourceCoords = getSourceCoords(this.aShape);
+    const userCode = `
+      void main() {
+        int index = int(gl_GlobalInvocationID.x);
+        ${dtype} resRC = getOutputCoords();
+        setOutput(index, getA(${sourceCoords}));
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/kernels/gather_webgpu.ts:52]: Do you need to do the out of bound checking?

**Review by haoyunfeix — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/4791)* status: 0 of 1 approvals obtained (waiting on @qjia7)

___
*[tfjs-backend-webgpu/src/kernels/gather_webgpu.ts, line 42 at r1](https://reviewable.io/reviews/tensorflow/tfjs/4791#-MVGNYSOMOilMmZthg3i-r1-42:-MVRsQci8L6fYk7UCzXd:b-896fix) ([raw file](https://github.com/tensorflow/tfjs/blob/433a31c5c334ae2c7581b822a69498ebc2a128b0/tfjs-backend-webgpu/src/kernels/gather_webgpu.ts#L42)):*
<details><summary><i>Previously, qjia7 (Jiajia Qin) wrote…</i></summary><blockquote>

`this.workPerThread` is not used in your shader. But you define it as a non-one value.
</blockquote></details>

Done.
___
*[tfjs-backend-webgpu/src/kernels/gather_webgpu.ts, line 52 at r1](https://reviewable.io/reviews/tensorflow/tfjs/4791#-MVGNYSOMOilMmZthg3i-r1-52:-MVRsS7X7DRVbjvGlFyV:b-896fix) ([raw file](https://github.com/tensorflow/tfjs/blob/433a31c5c334ae2c7581b822a69498ebc2a128b0/tfjs-backend-webgpu/src/kernels/gather_webgpu.ts#L52)):*
<details><summary><i>Previously, qjia7 (Jiajia Qin) wrote…</i></summary><blockquote>

Do you need to do the out of bound checking?
</blockquote></details>

Done.

File: tfjs-backend-webgpu/src/kernels/GatherV2.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+  toDispose.push(flattenIndex);
+
+  const flattenOutputShape = [
+    shapeInfo.batchSize, shapeInfo.outerSize, indicesSize / shapeInfo.batchSize,
+    shapeInfo.sliceSize
+  ];
+
+  if (backend.shouldExecuteOnCPU([x, indices]) || x.dtype === 'string') {
+    const indicesBufferInfo = backend.tensorMap.get(flattenIndex.dataId);
+    const indicesValues = indicesBufferInfo.values as TypedArray;
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/kernels/GatherV2.ts:69]: What if `x.dtype = 'string` but `indicesBufferInfo.values = null`?

**Review by haoyunfeix — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/4791)* status: 0 of 1 approvals obtained (waiting on @qjia7)

___
*[tfjs-backend-webgpu/src/kernels/GatherV2.ts, line 69 at r2](https://reviewable.io/reviews/tensorflow/tfjs/4791#-MVGNYSOMOilMmZthg3h-r2-69:-MVUtqyy5i05mEfgh_Ap:b-hsg0ek) ([raw file](https://github.com/tensorflow/tfjs/blob/f0f7e4a6ea66ddeb139a9f7f5133c8acc39f3133/tfjs-backend-webgpu/src/kernels/GatherV2.ts#L69)):*
<details><summary><i>Previously, qjia7 (Jiajia Qin) wrote…</i></summary><blockquote>

What if `x.dtype = 'string` but `indicesBufferInfo.values = null`?
</blockquote></details>

Just remove this pattern, not only indicesBuffer but also xBuffer should be both on CPU side, this is covered by backend.shouldExecuteOnCPU() method.
___
*[tfjs-backend-webgpu/src/kernels/gather_webgpu.ts, line 42 at r1](https://reviewable.io/reviews/tensorflow/tfjs/4791#-MVGNYSOMOilMmZthg3i-r1-42:-MVUuqkg01YiyP4m2you:b-e7bx87) ([raw file](https://github.com/tensorflow/tfjs/blob/433a31c5c334ae2c7581b822a69498ebc2a128b0/tfjs-backend-webgpu/src/kernels/gather_webgpu.ts#L42)):*
<details><summary><i>Previously, haoyunfeix (Hao Yunfei) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Just remove workPerThread
___
*[tfjs-backend-webgpu/src/kernels/gather_webgpu.ts, line 52 at r1](https://reviewable.io/reviews/tensorflow/tfjs/4791#-MVGNYSOMOilMmZthg3i-r1-52:-MVUuXRU60oG31v0q0w2:b-81zx6z) ([raw file](https://github.com/tensorflow/tfjs/blob/433a31c5c334ae2c7581b822a69498ebc2a128b0/tfjs-backend-webgpu/src/kernels/gather_webgpu.ts#L52)):*
<details><summary><i>Previously, haoyunfeix (Hao Yunfei) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

make it simple to check bounding with index and  size

File: tfjs-backend-webgpu/src/kernels/gather_webgpu.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+          setOutput(index, getA(${sourceCoords}));
+        }
+      }
+    `;
+    return userCode;
+  }
+}
+
+function getSourceCoords(aShape: number[]): string {
+  const currentCoords = ['resRC.x', 'resRC.y', 'resRC.z', 'resRC.w'];
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/kernels/gather_webgpu.ts:62]: I am wondering if the output is always flattened into rank 4 tensors. If yes, you can directly use `ivec4 resRC = getOutputCoords();` in L51 and remove L31, L38, L45. If no, you can't assume it as rank 4 here and use resRC.y/resRC.z/resRC.w

**Review by qjia7 — APPROVED** [review]
LGTM with a nit.
@lina128 @kainino0x take a look, thanks.

**Review by haoyunfeix — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/4791)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @kainino0x, @lina128, and @qjia7)

___
*[tfjs-backend-webgpu/src/kernels/gather_webgpu.ts, line 62 at r3](https://reviewable.io/reviews/tensorflow/tfjs/4791#-MVGNYSOMOilMmZthg3i-r3-62:-MV_K6bQA8tEEpFVllb4:b-dtchfo) ([raw file](https://github.com/tensorflow/tfjs/blob/cd46ed93d325c6196e5912116ab70b2d17710ef9/tfjs-backend-webgpu/src/kernels/gather_webgpu.ts#L62)):*
<details><summary><i>Previously, qjia7 (Jiajia Qin) wrote…</i></summary><blockquote>

 I am wondering if the output is always flattened into rank 4 tensors. If yes, you can directly use `ivec4 resRC = getOutputCoords();` in L51 and remove L31, L38, L45. If no, you can't assume it as rank 4 here and use resRC.y/resRC.z/resRC.w
</blockquote></details>

Thanks for your advice, done.

**Review by lina128 — APPROVED** [review]
LGTM

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/4791#-:-MVboYFr3W9Y7Lu_YMNG:bb6ywmx)* status: :shipit: complete! 2 of 1 approvals obtained (waiting on @kainino0x and @qjia7)

**Pull request merged**